transaction/beef.go

```diff
@@ -803,115 +800,214 @@ func (b *Beef) Verify(ctx context.Context, chainTracker chaintracker.ChainTracke
+					needsValidation = append(needsValidation, beefTx)
+				}
 			} else {
-				queue = append(queue, beefTx)
+				// Invalid bump index - treat as needing validation
+				needsValidation = append(needsValidation, beefTx)
+			}
+		case RawTx:
+			// RawTx is valid if it appears in a BUMP
+			if txidsInBumps[txid] {
+				validTxids[txid] = true
+				hasProof = append(hasProof, beefTx)
+			} else if beefTx.Transaction != nil {
+				// Check if all inputs are available
+				hasMissing := false
+				for _, input := range beefTx.Transaction.Inputs {
+					if _, exists := b.Transactions[input.SourceTXID.String()]; !exists {
+						missingInputs[input.SourceTXID.String()] = true
+						hasMissing = true
+					}
+				}
+				if hasMissing {
+					withMissingInputs = append(withMissingInputs, beefTx)
+				} else {
+					needsValidation = append(needsValidation, beefTx)
+				}
 			}
 		}
 	}
 
-	// Try to validate any transactions whose inputs are now known
-	oldLen := -1
-	for oldLen != len(queue) {
-		oldLen = len(queue)
-		newQueue := make([]*BeefTx, 0, len(queue))
-		for _, beefTx := range queue {
+	// Iteratively validate transactions that depend on other transactions
+	for len(needsValidation) > 0 {
+		progress := false
+		var stillNeedsValidation []*BeefTx
+
+		for _, beefTx := range needsValidation {
+			// Check if all inputs are valid
+			allInputsValid := true
 			if beefTx.Transaction != nil {
-				allInputsValid := true
-				for _, in := range beefTx.Transaction.Inputs {
-					if !validTxids[in.SourceTXID.String()] {
+				for _, input := range beefTx.Transaction.Inputs {
+					if !validTxids[input.SourceTXID.String()] {
 						allInputsValid = false
 						break
 					}
 				}
-				if allInputsValid {
-					validTxids[beefTx.Transaction.TxID().String()] = true
-					res.Valid = append(res.Valid, beefTx.Transaction.TxID().String())
-				} else {
-					newQueue = append(newQueue, beefTx)
+			}
+
+			if allInputsValid {
+				txid := beefTx.Transaction.TxID().String()
+				validTxids[txid] = true
+				hasProof = append(hasProof, beefTx)
+				progress = true
+			} else {
+				stillNeedsValidation = append(stillNeedsValidation, beefTx)
+			}
+		}
+
+		needsValidation = stillNeedsValidation
+		if !progress {
+			// No progress made - remaining transactions are not valid
+			for _, beefTx := range needsValidation {
+				if beefTx.Transaction != nil {
+					result.NotValid = append(result.NotValid, beefTx.Transaction.TxID().String())
 				}
 			}
+			break
+		}
+	}
+
+	// Populate result lists
+	// Add transactions with missing inputs
+	for _, beefTx := range withMissingInputs {
+		if beefTx.Transaction != nil {
+			txid := beefTx.Transaction.TxID().String()
+			result.WithMissingInputs = append(result.WithMissingInputs, txid)
 		}
-		queue = newQueue
 	}
 
-	// Now, whatever is left in queue is not valid
-	for _, beefTx := range queue {
+	// Add txid-only transactions
+	for _, beefTx := range txidOnly {
```

Comment:
Since the length of txidonly is known, I would preallocate memory for `result.TxidOnly` slice.
```suggestion
	result.TxidOnly = make([]string, 0, len(txidOnly))
	for _, beefTx := range txidOnly {
```